cEP-0024.md

```diff
@@ -0,0 +1,226 @@
+# Implement GitMate automations and plugins for coala
+
+| Metadata |                                                       |
+| -------- | ----------------------------------------------------- |
+| cEP      | 24                                                    |
+| Version  | 1.0                                                   |
+| Title    | Implement Gitmate automations and plugins for coala   |
+| Authors  | Vamshi Krishna Bommerla <mailto:bommvams72@gmail.com> |
+| Status   | Proposed                                              |
+| Type     | Process                                               |
+
+## Abstract
+
+This cEP describes the improvements to the coala workflow and the necessary
+automation in [GitMate](https://gitmate.io/) to realise it, as a part of the
+[GSoC project](https://summerofcode.withgoogle.com/projects/#5344316227059712).
+
+## Introduction
+
+[GitMate](https://gitmate.io/) is an automation tool for developers. While it
+works well, there are a number of missing features which would save time of
+coala community, for example automatically close PRs opened on master of forks.
+Many of the plugins in GitMate are general plugins, which are required by every
+organization, but coala needs more plugins which support coala development
+workflow. This project adds plugins to auto-review, auto-reject, auto-label
+PRs, assign, unassign issues, squash commits.
+
+## Changes in coala workflow
+
+### Use increasing difficulty level order to assign issues
+
+Current assign process in coala is to check team membership of a user,
+difficulty level of the issue and assign issue to user if eligible. To
+implement current assign process in GitMate, IGitt should support teams,
+but it is difficult to implement, because there are no teams in GitLab.
+
+Considering teams as groups in GitLab is a possible solution for IGitt
+implementation, but requires extra sync tools to sync GitHub team and GitLab
+group memberships, which nobody has built yet.
+
+Team membership on GitHub is not publicly accessible information, so any system
+which wants to implement the current 'assign' system needs to have a master
+org key, which isn't good.
+
+Best possible solution would be using a increasing difficulty level order,
+so that a user can be assigned to an issue only if user has solved issues of
+lower difficulty level before requesting to get assigned to a higher difficulty
+level issue.
+
+#### Drawbacks
+
+If user completed the lower difficulty level issues in GitHub and want to
+work on a higher difficulty level issue in GitLab, he has to repeat the
+newcomer process in GitLab.
+
+If a user is worthy enough to work on the higher level difficulty issues
+directly, then that user needs to be manually assigned by the maintainers.
```

Comment:
the important point to make here is that

- currently maintainers can put the 'special' newcomer into the developers group quickly and no manual assignment is needed after that.

- if using increasing order, the maintainer will need to keep doing manual assignment until the 'special' newcomer has completed a `low`.

In reality those two are not very different.  Both require at least one manual intervention by maintainers.  And both are unlikely to involve two manual interventions by maintainers as pushing people into developers early is a very rare occurrence currently, and if using difficulty order we would be unlikely to assign multiple issues to even a 'special' newcomer until they have finished at least one -- if they are special, they should be able to get the first 'low' done quickly and the problem disappears.

IMO your proposed model is actually better for these 'special' newcomers, as they can be assigned a low or medium by a maintainer, and if they do it then the one minor manual intervention was all that was required and 'developer' status doesnt have to be granted out of process.

Socially it is more acceptable to assign these special newcomers an issue and they complete it, rather than granting them developer status when they havent really earnt it yet.